packages/vm/src/context.rs

```diff
@@ -20,30 +20,79 @@ use cosmwasm_std::KV;
 #[cfg(feature = "iterator")]
 use crate::errors::{make_iterator_does_not_exist, FfiResult};
 use crate::errors::{make_uninitialized_context_data, VmResult};
+#[cfg(feature = "iterator")]
+use crate::traits::StorageIteratorItem;
 use crate::traits::{Querier, Storage};
 
 /** context data **/
 
+#[derive(Clone, PartialEq, Debug, Default)]
+pub struct GasState {
+    /// Gas limit for the computation.
+    gas_limit: u64,
+    /// Tracking the gas used in the cosmos SDK, in cosmwasm units.
+    #[allow(unused)]
+    externally_used_gas: u64,
+}
+
+impl GasState {
+    fn with_limit(gas_limit: u64) -> Self {
+        Self {
+            gas_limit,
+            externally_used_gas: 0,
+        }
+    }
+
+    #[allow(unused)]
+    fn use_gas(&mut self, amount: u64) {
+        self.externally_used_gas += amount;
+    }
+
+    pub(crate) fn set_gas_limit(&mut self, gas_limit: u64) {
+        self.gas_limit = gas_limit;
+    }
+
+    /// Get the amount of gas units still left for the rest of the calculation.
+    ///
+    /// We need the amount of gas used in wasmer since it is not tracked inside this object.
+    fn get_gas_left(&self, wasmer_used_gas: u64) -> u64 {
```

Comment:
Ah, I was right somehow that we were double-subtracting gas.
We removed external gas from the limit, then counted that against internal usage later.

Thanks for testing this and pulling out this calculation

Reply:
yeah, i'm super happy we caught this now. thanks for making me do it :)